File: Trie/440. K-th Smallest in Lexicographical Order.cpp
/*
440. K-th Smallest in Lexicographical Order
*/

// TC:O(lgn * lgn) SC:O(lgn)
class Solution {
    using LL = long long;
public:
    int findKthNumber(int n, int k) {
        return FindKthNumberBeginWith(0, n, k);
    }

    // ~= 10^2
    // return the Lexicographically k-th element that begins with prefix
    // if k==0, return prefix itself
    int FindKthNumberBeginWith(int prefix, int n, int k){
        if(k == 0) return prefix;
        int start = (prefix == 0 ? 1 : 0);

        for(int i = start; i <= 9; i++){
            // 1 表示目前的數字就是結尾的狀況
            int count = 1 + TotalNumbersBeginWith(prefix*10+i, n);

            if(count < k){
                k -= count;
            }
            else {
                return FindKthNumberBeginWith(prefix*10+i, n, k-1);
            }
        }

        return -1;
    }

    // return how many numbers that begin with prefix (and <= n)
    // excluding prefix itself
    int TotalNumbersBeginWith(int prefix, int n){
        
        LL exp = 10;
        LL count = 0;
        while(1){
            LL lower = prefix*exp;
            LL upper = prefix*exp + exp - 1;

            if(lower > n) break;
            // 5050, 4050 -> 101
            else if(lower <= n && upper >= n){
                count += (n-lower+1);
                break;
            }
            else {
                count += exp;
            }
            exp *= 10;
        }

        return count;
    }
};

File: Trie/440. K-th Smallest in Lexicographical Order.py
# 優化前: TC:O(D * lgn * lgn) SC:O(1) 優化後: TC:O(D * lgn) SC:O(1)
class Solution:
    def findKthNumber(self, n: int, k: int) -> int:
        x = n // 10
        subtree_size = 0
        pow10 = 1
        while x > 0:
            subtree_size = subtree_size * 10 + 1
            x //= 10
            pow10 *= 10
        def count_subtree_size(node:int) -> int:
            left, right = node * pow10, (node + 1) * pow10
            size = subtree_size
            if left <= n:
                size += min(n + 1, right) - left
            return size
        node = 1
        k -= 1
        while k > 0:
            size = count_subtree_size(node)
            if size <= k:
                k -= size
                node += 1
            else:
                subtree_size //= 10
                pow10 //= 10
                node *= 10
                k -= 1
        return node

# 優化前
class Solution:
    def findKthNumber(self, n: int, k: int) -> int:
        def count_subtree_size(node:int) -> int:
            left, right = node, node + 1
            size = 0
            while left <= n:
                size += min(n + 1, right) - left
                left *= 10
                right *= 10
            return size
        node = 1
        k -= 1
        while k > 0:
            size = count_subtree_size(node)
            if size <= k:
                k -= size
                node += 1
            else:
                node *= 10
                k -= 1
        return node
